File: backend/app/models/monitor.py
import enum
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import BigInteger, Boolean, DateTime, Enum, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class AuditTargetType(str, enum.Enum):
    user = "user"
    role = "role"
    resource = "resource"
    system = "system"


class AuditOperation(str, enum.Enum):
    create = "create"
    read = "read"
    update = "update"
    delete = "delete"
    login_success = "login_success"
    login_failure = "login_failure"
    permission_check = "permission_check"
    role_assign = "role_assign"
    role_revoke = "role_revoke"
    password_reset = "password_reset"
    suspend = "suspend"
    restore = "restore"
    block = "block"
    request_submit = "request_submit"
    request_approve = "request_approve"
    request_reject = "request_reject"
    request_withdraw = "request_withdraw"


class AuditModule(str, enum.Enum):
    identity = "identity"
    access = "access"
    monitor = "monitor"
    reports = "reports"
    auth = "auth"


class AuditResult(str, enum.Enum):
    success = "success"
    failure = "failure"
    denied = "denied"


class OutboxStatus(str, enum.Enum):
    pending = "pending"
    published = "published"
    failed = "failed"


class AlertConditionType(str, enum.Enum):
    threshold = "threshold"
    pattern = "pattern"
    anomaly = "anomaly"


class AlertSeverity(str, enum.Enum):
    info = "info"
    low = "low"
    medium = "medium"
    high = "high"
    critical = "critical"


class AlertDataSource(str, enum.Enum):
    postgres = "postgres"
    elasticsearch = "elasticsearch"


class AlertStatus(str, enum.Enum):
    new = "new"
    acknowledged = "acknowledged"
    resolved = "resolved"
    false_positive = "false_positive"


class NotificationChannelType(str, enum.Enum):
    email = "email"
    webhook = "webhook"
    log = "log"
    kafka = "kafka"


class AuditLog(Base):
    __tablename__ = "audit_log"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    event_id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), unique=True, nullable=False, default=uuid4)
    timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), index=True
    )
    actor_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id", ondelete="SET NULL"), nullable=True
    )
    actor_username: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    target_type: Mapped[AuditTargetType] = mapped_column(
        Enum(AuditTargetType, name="audit_target_type"), nullable=False
    )
    target_id: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    operation: Mapped[AuditOperation] = mapped_column(
        Enum(AuditOperation, name="audit_operation"), nullable=False
    )
    module: Mapped[AuditModule] = mapped_column(
        Enum(AuditModule, name="audit_module"), nullable=False
    )
    result: Mapped[AuditResult] = mapped_column(
        Enum(AuditResult, name="audit_result"), nullable=False
    )
    ip_address: Mapped[str | None] = mapped_column(String(45), nullable=True)
    user_agent: Mapped[str | None] = mapped_column(String(500), nullable=True)
    details: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    correlation_id: Mapped[UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    published_to_kafka: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class OutboxEvent(Base):
    __tablename__ = "outbox_events"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    audit_log_id: Mapped[int] = mapped_column(
        BigInteger, ForeignKey("audit_log.id", ondelete="CASCADE"), nullable=False
    )
    topic: Mapped[str] = mapped_column(String(100), nullable=False)
    payload: Mapped[dict] = mapped_column(JSONB, nullable=False)
    status: Mapped[OutboxStatus] = mapped_column(
        Enum(OutboxStatus, name="outbox_status"), nullable=False, default=OutboxStatus.pending
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    published_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    attempts: Mapped[int] = mapped_column(Integer, nullable=False, default=0)


class AlertRule(Base):
    __tablename__ = "alert_rules"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False, default="")
    condition_type: Mapped[AlertConditionType] = mapped_column(
        Enum(AlertConditionType, name="alert_condition_type"), nullable=False
    )
    condition_config: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    severity: Mapped[AlertSeverity] = mapped_column(
        Enum(AlertSeverity, name="alert_severity"), nullable=False
    )
    is_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    cooldown_seconds: Mapped[int] = mapped_column(Integer, nullable=False, default=300)
    data_source: Mapped[AlertDataSource] = mapped_column(
        Enum(AlertDataSource, name="alert_data_source"), nullable=False
    )

    alerts: Mapped[list["Alert"]] = relationship("Alert", back_populates="rule")


class Alert(Base):
    __tablename__ = "alerts"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    rule_id: Mapped[UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("alert_rules.id"), nullable=False
    )
    triggered_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    subject_user_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id", ondelete="SET NULL"), nullable=True
    )
    severity: Mapped[AlertSeverity] = mapped_column(
        Enum(AlertSeverity, name="alert_severity"), nullable=False
    )
    status: Mapped[AlertStatus] = mapped_column(
        Enum(AlertStatus, name="alert_status"), nullable=False, default=AlertStatus.new
    )
    correlation_id: Mapped[UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    details: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    acknowledged_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    acknowledged_by: Mapped[UUID | None] = mapped_column(PGUUID(as_uuid=True), nullable=True)
    resolution_comment: Mapped[str | None] = mapped_column(Text, nullable=True)

    rule: Mapped["AlertRule"] = relationship("AlertRule", back_populates="alerts")


class NotificationChannel(Base):
    __tablename__ = "notification_channels"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    type: Mapped[NotificationChannelType] = mapped_column(
        Enum(NotificationChannelType, name="notification_channel_type"), nullable=False
    )
    config: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    is_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

File: frontend/src/api/monitor.ts
import { api } from "./client";

export type AuditTargetType = "user" | "role" | "resource" | "system";
export type AuditOperation =
  | "create" | "read" | "update" | "delete"
  | "login_success" | "login_failure" | "permission_check"
  | "role_assign" | "role_revoke" | "password_reset"
  | "suspend" | "restore" | "block"
  | "request_submit" | "request_approve" | "request_reject" | "request_withdraw";
export type AuditModule = "identity" | "access" | "monitor" | "reports" | "auth";
export type AuditResult = "success" | "failure" | "denied";
export type AlertSeverity = "info" | "low" | "medium" | "high" | "critical";
export type AlertStatus = "new" | "acknowledged" | "resolved" | "false_positive";
export type AlertDataSource = "postgres" | "elasticsearch";
export type AlertConditionType = "threshold" | "pattern" | "anomaly";
export type NotificationChannelType = "email" | "webhook" | "log" | "kafka";

export interface AuditLogEntry {
  id: number;
  event_id: string;
  timestamp: string;
  actor_id: string | null;
  actor_username: string;
  target_type: AuditTargetType;
  target_id: string;
  operation: AuditOperation;
  module: AuditModule;
  result: AuditResult;
  ip_address: string | null;
  user_agent: string | null;
  details: Record<string, unknown> | null;
  correlation_id: string | null;
  published_to_kafka: boolean;
}

export interface AuditLogListResponse {
  items: AuditLogEntry[];
  total: number;
  page: number;
  page_size: number;
}

export interface AlertRule {
  id: string;
  code: string;
  name: string;
  description: string;
  condition_type: AlertConditionType;
  condition_config: Record<string, unknown>;
  severity: AlertSeverity;
  is_enabled: boolean;
  cooldown_seconds: number;
  data_source: AlertDataSource;
}

export interface AlertRulesListResponse {
  items: AlertRule[];
  total: number;
}

export interface Alert {
  id: string;
  rule_id: string;
  triggered_at: string;
  subject_user_id: string | null;
  severity: AlertSeverity;
  status: AlertStatus;
  correlation_id: string | null;
  details: Record<string, unknown>;
  acknowledged_at: string | null;
  acknowledged_by: string | null;
  resolution_comment: string | null;
  rule?: AlertRule;
}

export interface AlertsListResponse {
  items: Alert[];
  total: number;
  page: number;
  page_size: number;
}

export interface NotificationChannel {
  id: string;
  code: string;
  type: NotificationChannelType;
  config: Record<string, unknown>;
  is_enabled: boolean;
}

export interface DashboardMetrics {
  total_events_today: number;
  failed_logins_today: number;
  active_alerts: number;
  critical_alerts: number;
  events_by_module: Record<string, number>;
  events_by_result: Record<string, number>;
}

export const monitorApi = {
  getDashboard: () => api.get<DashboardMetrics>("/monitor/dashboard"),

  listAudit: (params: {
    actor_username?: string;
    operation?: string;
    module?: string;
    result?: string;
    date_from?: string;
    date_to?: string;
    page?: number;
    page_size?: number;
  }) => api.get<AuditLogListResponse>("/monitor/audit", { params }),

  getAuditEntry: (event_id: string) =>
    api.get<AuditLogEntry>(`/monitor/audit/${event_id}`),

  exportAudit: (params: { date_from?: string; date_to?: string; fmt?: string }) =>
    api.get("/monitor/audit/export", { params, responseType: "blob" }),

  listRules: () => api.get<AlertRulesListResponse>("/monitor/rules"),

  createRule: (body: Omit<AlertRule, "id">) =>
    api.post<AlertRule>("/monitor/rules", body),

  updateRule: (id: string, body: Partial<AlertRule>) =>
    api.patch<AlertRule>(`/monitor/rules/${id}`, body),

  toggleRule: (id: string) =>
    api.post<AlertRule>(`/monitor/rules/${id}/toggle`),

  testRule: (id: string) =>
    api.post<{ rule_code: string; matched: boolean; details: Record<string, unknown> }>(
      `/monitor/rules/${id}/test`
    ),

  listAlerts: (params: {
    status?: string;
    severity?: string;
    page?: number;
    page_size?: number;
  }) => api.get<AlertsListResponse>("/monitor/alerts", { params }),

  getAlert: (id: string) => api.get<Alert>(`/monitor/alerts/${id}`),

  acknowledgeAlert: (id: string, comment?: string) =>
    api.post<Alert>(`/monitor/alerts/${id}/acknowledge`, { comment }),

  resolveAlert: (id: string, comment?: string) =>
    api.post<Alert>(`/monitor/alerts/${id}/resolve`, { comment }),

  markFalsePositive: (id: string, comment?: string) =>
    api.post<Alert>(`/monitor/alerts/${id}/false-positive`, { comment }),

  listChannels: () => api.get<NotificationChannel[]>("/monitor/channels"),

  createChannel: (body: Omit<NotificationChannel, "id">) =>
    api.post<NotificationChannel>("/monitor/channels", body),

  updateChannel: (id: string, body: { config?: Record<string, unknown>; is_enabled?: boolean }) =>
    api.patch<NotificationChannel>(`/monitor/channels/${id}`, body),

  getKibanaToken: () =>
    api.get<{ embed_url: string; token: string | null }>("/monitor/kibana-token"),
};
